shared/src/main/scala/mlscript/Typer.scala

```diff
@@ -20,60 +20,94 @@ class Typer(var dbg: Boolean, var verbose: Bool, var explainErrors: Bool) extend
   type Raise = Diagnostic => Unit
   type Binding = Str -> TypeScheme
   type Bindings = Map[Str, TypeScheme]
-  
+
   case class TypeDef(
     kind: TypeDefKind,
     nme: TypeName,
-    tparams: List[TypeName],
-    targs: List[TypeVariable],
-    body: Type,
+    tparamsargs: List[(TypeName, TypeVariable)],
+    tvars: Iterable[TypeVariable], // "implicit" type variables. instantiate every time a `TypeRef` is expanded
+    bodyTy: SimpleType,
     mthDecls: List[MethodDef[Right[Term, Type]]],
     mthDefs: List[MethodDef[Left[Term, Type]]],
     baseClasses: Set[Var],
     toLoc: Opt[Loc],
   ) {
     def allBaseClasses(ctx: Ctx)(implicit traversed: Set[Var]): Set[Var] =
-      baseClasses.map(v => Var(v.name.toList.mapHead(_.toLower).mkString)) ++
+      baseClasses.map(v => Var(v.name.decapitalize)) ++
         baseClasses.iterator.filterNot(traversed).flatMap(v =>
           ctx.tyDefs.get(v.name).fold(Set.empty[Var])(_.allBaseClasses(ctx)(traversed + v)))
-    val targsMap: Map[Str, TypeVariable] = tparams.map(_.name).zip(targs).toMap
+    val tparams: List[TypeName] = tparamsargs.map(_._1)
+    val targs: List[TypeVariable] = tparamsargs.map(_._2)
+    def thisTy(implicit prov: TypeProvenance): TypeRef = TypeRef(nme, targs)(prov)
+    def wrapMethod(bodyTy: PolymorphicType)(implicit prov: TypeProvenance): MethodType =
+      MethodType(bodyTy.level, FunctionType(singleTup(thisTy), bodyTy.body)(prov), nme)
+  }
+
+  private case class MethodDefs(
+    tn: TypeName,
+    prts: List[MethodDefs],
+    decls: Map[Str, MethodType],
+    defns: Map[Str, MethodType],
+  ) {
+    private def &(that: MethodDefs)(tn: TypeName, newdefns: Set[Str])(implicit raise: Raise): MethodDefs =
+      MethodDefs(
+        tn,
+        this.prts ::: that.prts,
+        mergeMap(this.decls, that.decls)(_ & _),
+        (this.defns.iterator ++ that.defns.iterator).toSeq.groupMap(_._1)(_._2).flatMap {
+          case _ -> Nil => die
+          case mn -> (defn :: Nil) => S(mn -> defn)
+          case mn -> defns if newdefns(mn) => N
+          case mn -> defns =>
+            implicit val prov: TypeProvenance = TypeProvenance(tn.toLoc, "inherited method declaration")
+            S(mn -> MethodType(defns.head.level,
+              err(msg"A method definition must be given when inheriting multiple method definitions" -> tn.toLoc
+                :: msg"Definitions of method $mn inherited from:" -> N
+                :: defns.iterator.map(mt => msg"• ${mt.prts.head}" -> mt.prov.loco).toList),
+              tn :: Nil
+            ))
+        }
+      )
+    private def ++(that: MethodDefs): MethodDefs = {
+      require(prts.isEmpty)
+      MethodDefs(
+        that.tn,
+        that.prts,
+        this.decls ++ that.decls,
+        this.defns ++ that.defns
+      )
+    }
+    private def addprt(prt: TypeName): MethodDefs =
```

Comment:
Method naming consistency:
```suggestion
    private def addParent(prt: TypeName): MethodDefs =
```

Reply:
Addressed in f48c96fcda9dfbf2f7633ccfec76d4387d9aa354